certbot_dns_duckdns/duckdns/client.py

```diff
@@ -63,10 +64,19 @@ def __get_validated_root_domain__(domain):
         # get the root domain with the first subdomain
         domain_parts = domain.split(".")
         root_domain = ".".join(domain_parts[-3:])
+
+        # if the root domain is not a duckdns.org domain, check if it delegates the acme challenge and resolve the resulting CNAME.
+        # see delegated acme challenge https://letsencrypt.org/docs/challenge-types/#dns-01-challenge
+        if VALID_DUCKDNS_DOMAIN_REGEX.match(root_domain) == None:
+            result = resolver.resolve("_acme-challenge." + domain, 'A')
+            root_domain = result.canonical_name.to_text().rstrip('.') # Nobody84: In my testing the canonical name end with a '.' like 'example.duckdns.org.' 
```

Comment:
We should not make the resolving of the CNAME to the domain here.

This has 2 reasons:
1. The DuckDNS client is only responsible for interacting with the DuckDNS API and should not resolve DNS records as part of the DNS challange. A suitable place would be the [Authenticator class](https://github.com/infinityofspace/certbot_dns_duckdns/blob/0c594a7e55ac720166fd111e268dc361630c7b62/certbot_dns_duckdns/cert/client.py#L75) before setting the TXT entry, which exists exactly for this responsibility.
2. Because of these responsibilities, the Authenticator also takes over the resolution of current TXT entries and uses them depending on the situation. This would cause the TXT entry to be selected incorrectly with your current implementation. Therefore your CNAME resolution implementation must be done exactly before the request of the old TXT entry.

Another problem is that you retrieve the A record of the domain to get the final domain from the CNAME. This works fine, but not as soon as only an IPv6 address (AAAA record) is stored instead of an IPv4 address (A record). Thus you would have to query the AAAA if the A record is missing.

Reply:
Hi,

I totally missed the AAAA record, because I do not use it in my scenario, my fault. 
I will look at the [Authenticator class](https://github.com/infinityofspace/certbot_dns_duckdns/blob/0c594a7e55ac720166fd111e268dc361630c7b62/certbot_dns_duckdns/cert/client.py#L75) and move the CNAME resolution there.